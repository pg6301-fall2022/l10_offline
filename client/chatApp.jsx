import React, {useState} from "react";


export function ChatApp({ messages, onNewMessage }){
    const [message, setMessage] = useState("");

    function handleSubmit(e){
        e.preventDefault();
        onNewMessage(message);
        setMessage("");
    }

    return(
        <>
            <header> PG6301 - HK - Chat </header>
            <main>
                {
                    messages.map(( { message, user}, index ) => (
                        <div key={index}>
                            <strong> {user}: </strong> {message}
                        </div>
                    ))
                }
            </main>
            <footer>
                <form onSubmit={handleSubmit}>
                    <input
                        autoFocus={true}
                        value={message}
                        onChange={(e) => setMessage(e.target.value)}
                    />
                    <button> Send </button>
                </form>
            </footer>
        </>
    );
}