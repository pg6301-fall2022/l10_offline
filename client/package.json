{
  "name": "client",
  "version": "1.0.0",
  "description": "",
  "scripts": {
    "test": "jest",
    "dev": "parcel watch index.html",
    "build": "npm install --include=dev && parcel build index.html"
  },
  "keywords": [],
  "author": "",
  "license": "ISC",
  "dependencies": {
    "@babel/preset-env": "^7.20.2",
    "@babel/preset-react": "^7.18.6",
    "babel-jest": "^29.2.2",
    "jest": "^29.2.2",
    "jest-environment-jsdom": "^29.2.2",
    "jsdom": "^20.0.2",
    "react": "^18.2.0",
    "react-dom": "^18.2.0",
    "ws": "^8.11.0"
  },
  "devDependencies": {
    "parcel": "^2.7.0",
    "process": "^0.11.10"
  },
  "jest": {
    "testEnvironment": "jsdom"
  },
  "babel": {
    "presets": [
      "@babel/preset-env",
      "@babel/preset-react"
    ]
  }
}
