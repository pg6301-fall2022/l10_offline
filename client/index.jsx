import React, {useEffect, useState} from "react";
import { createRoot } from "react-dom/client";
import { ChatApp } from "./chatApp";

const initialMessages = [
    {
        user: "User1",
        message: "Oj there from user1",
    } ,
    {
        user: "User2",
        message: "Hello there from user2",
    },
];

function UserRegistration( { onUsername } ){
    const [username, setUsername] = useState();

    function handleSubmit(e){
        e.preventDefault();
        onUsername(username);
    }

    return(
        <form onSubmit={handleSubmit}>
            <label>
                Username:
                <input value={username} onChange={(e) => setUsername(e.target.value)}/>
            </label>
            <button> Submit </button>
        </form>
    );
}

function Application(){
    const [ messages, setMessages ] = useState(initialMessages);
    const [ user, setUser ] = useState();
    const [ ws, setWs ] = useState();

    useEffect(() => {
       const ws = new WebSocket("ws://" + window.location.host);

       ws.onopen = (event) => {
           console.log(" Web socket open!")
       }
       ws.onmessage = (event) => {
           console.log("New message ", event);
       }

       setWs(ws);
    }, []);

    function onNewMessage(message){
        setMessages((messages) => [...messages, {message, user}]);
        ws.send("A new message " + message);
    }

    if(!user){
      return <UserRegistration onUsername={setUser} />
    }

    return(
        <ChatApp messages={messages} onNewMessage={onNewMessage} />
    );
}

const root = createRoot(document.getElementById("app"));
root.render(<Application />);