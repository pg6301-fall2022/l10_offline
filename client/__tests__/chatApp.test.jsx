import React from "react";
import { createRoot } from "react-dom/client";
import { ChatApp } from "../chatApp";
import { act, Simulate } from "react-dom/test-utils";



describe("chat test suite", () => {
   it("shows messages", () =>{
        const testMess = [
            { user: "Test1", message: "Yep"},
            { user: "Test2", message: "Nope"},
        ];

        const element = document.createElement("div");
        const root = createRoot(element);

        act(() =>
            root.render(<ChatApp messages={testMess} /> )
        );

        expect(element.innerHTML).toMatchSnapshot();
   });

   it("adds new message", () => {
       const element = document.createElement("div");
       const root = createRoot(element);

       const onNewMessage = jest.fn();

       act(() =>
        root.render(<ChatApp messages={[]} onNewMessage={onNewMessage} />)
       );

       act(() =>
           Simulate.change(element.querySelector("footer input"), {
               target: {value: "something else"}
           })
       );

       act(() =>
            Simulate.submit(element.querySelector("footer form"))
       );

       expect(onNewMessage).toBeCalledWith("something else");

   });
});