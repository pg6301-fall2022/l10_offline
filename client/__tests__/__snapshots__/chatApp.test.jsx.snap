// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`chat test suite shows messages 1`] = `"<header> PG6301 - HK - Chat </header><main><div><strong> Test1: </strong> Yep</div><div><strong> Test2: </strong> Nope</div></main><footer><form><input value=""><button> Send </button></form></footer>"`;
